### ISSUE TITLE
Why is the public key 66 and not 64 hex characters long?

### ISSUE THREAD

**Issue by ilmoi** [rust-bitcoin/rust-secp256k1#319]
I'm new to crypto so pardon if this is a total noob question - but why is the public key 66 chacracters long and not 64? I was under the impression that public keys are [64 chars](https://davidederosa.com/basic-blockchain-programming/elliptic-curve-keys/#:~:text=Public%20keys%20are%2064%20bytes,is%20based%20on%20modular%20arithmetic.) (256 bit)?

And on a related note, I'm trying to convert a 64-byte keccak256 hash into an instance of a `PublicKey` and am getting an `InvalidPublicKey` error. Is that because I'm attempting something fundamentally incompatible, or am I massaging the input wrong? This is where my 64 vs 66 character observation came from.

My naive attempt:
```
let keccak_hash = "90a39c61834946715b38aa7eb6b99fb88a5df061299ae0813c97d8b5fdf3d15c";
let pk = PublicKey::from_str(keccak_hash).unwrap();
```

**Comment by TheBlueMatt** [comment]
Public keys are *points* (ie have an X and Y coordinate), not arbitrary byte strings. In "compressed" form, they are serialized as only the X coordinate and a bit to indicate if they are positive or negative (as there are only two possible Ys for each X on the curve). You cannot select an arbitrary string and assume it is a valid public key, though you *usually* can assume a securely random 32-byte string is a valid secret key (there are only a few examples where they are not - 0, a few very high numbers, etc, but if your string is really securely random, you should only see that once or twice before the universe burns up).

**Comment by ilmoi** [comment]
Thanks @TheBlueMatt. This made me realize I had to go do a lot more reading.

For the specific question in the title, if anyone else is wondering, the public key is 66 hex chars (33 bytes) because it's a compressed key (that also explains why they start with 02 and 03). An uncompressed key would actually be twice as long - 130 hex chars. That's how ethereum has it.

I found [this chapter](https://github.com/ethereumbook/ethereumbook/blob/develop/04keys-addresses.asciidoc) of the Mastering Ethereum book to be very accessible for someone without a cryptography background like myself. They explain the compressed vs uncompressed encodings and how public keys are converted to ethereum addresses.

Hope helps someone in the future.
